### ISSUE TITLE
Integration with logging

### ISSUE THREAD

**Issue by KyleRAnderson** [FRC2706/2706Dashboard#1]
Should have the option to view logs generated on the robot each match. 
The purpose would be for debugging both during the development stage and at competitions. Things like autonomous mode could really use the ability to quickly view the log output for their subsystem between matches.

Requirements
- Doesn't take up bandwidth during matches
- Accessible while the robot is off (pre-downloaded, stored on external storage, ...)

**Comment by ounsworth** [comment]
I love the idea of logs going to the DS so we don't need to ssh into the roborio to see what happened in the last match.
We do need to be a bit careful about how much bandwidth the logging is using so it's not competing with the camera stream during a match. Is it possible to do this "on demand" (when the user presses a "Fetch Logs" button?) @eandr127  ?

**Comment by KyleRAnderson** [comment]
Yeah, sorry Mike. You made this comment earlier on the other repository and I didn't copy your comments over when I moved the issues.

**Comment by ounsworth** [comment]
No worries, that's why I made the comment again :)

**Comment by georgetzavelas** [comment]
Just a few questions so I better understand the requirement.

What is the purpose of viewing the logs? Which logs are we viewing? If it is the info that is dumped to stdout/stderr? Is there a way to obtain that information after the robot has been powered off?

**Comment by KyleRAnderson** [comment]
@georgetzavelas I updated some of the requirements for this. This issue was kind of just made as a temporary ticket to which we would add more information once we knew exactly what we wanted out of this feature, a work in progress ticket kind of.

Basically, Ryan worked on some logging last year and we want those logs to be easily accessible at competitions and during debugging. The way we do this or specific requirements for this feature have not yet been decided.

Not sure whether or not it would be accessible after shutdown. That's another thing that we haven't investigated. We've been focusing on getting the other, more important features working in the dashboard before this one.

**Comment by ryanlarkin** [comment]
Currently the logging is printed to the robot's stdout and the driver station as log files, but is not saved at all on the roboRIO. It wouldn't take much work to save the logs to the robot, and would mean less bandwidth used. Logs could be accessed from any computer, but would be more difficult to get to in between matches, as the robot would need to be powered on and have ssh be used to retrieve the logs.

A 'download logs' button that Mike suggested could also work, as it would only get the logs as necessary, with the only drawback being that someone would actually have to remember to press the download button, which could be a job for the technician. This also shouldn't require too many code changes to implement.

Another possibility is to upload/download the logs when the robot is disabled to conserve bandwidth when it's needed.

**Comment by ounsworth** [comment]
Ooo! I like the idea of pushing the logs in disabledInit() and just append them to end of a giant long file on the DS.
(depending how much we log, like if we start logging raw encoder vals, then after pushing to the DS, we may need to clean up the file on the roboRIO so we don't run out of disk)

**Comment by KyleRAnderson** [comment]
Interesting, we'll look in to it.
I guess it would have to download at the beginning and end of matches then. I'm wondering how it would handle interruptions if the robot is turned off during the download. Depending on the size of the log files, as Mike says, it make take a long time.

**Comment by kevlam2706** [comment]
Does our logging infrastructure support the standard
Errors/Warnings/Info/Debug levels?

**Comment by KyleRAnderson** [comment]
Can you put an SD card into the RoboRIO? Or any other storage device? If so, we could store the logs there possibly.

**Comment by ryanlarkin** [comment]
It uses `java.util.logging`, so it has all logging levels.

Last year we used both our USB ports for driver cameras, but if there's another simple way to add a removable storage device to the roboRIO we could definitely use that.

**Comment by georgetzavelas** [comment]
How much local storage does the RoboRio have? 

Alternatively a really nice project would be to explore using Kafka to stream the logs https://kafka.apache.org/ along side the image stream. Cool technology that has widespread adoption.

**Comment by kevlam2706** [comment]
According to Screen Steps, FIRST is providing some kind of new dashboard system with the ability to stream and record telemetry from the robot. I haven’t looked too deeply into what this is or how it works, but someone should :) 

Kevin

**Comment by ounsworth** [comment]
Kevin, would you be willing to do some reading into that technology and see if it meets the two requirements that Kyle has posted in the top comment on this thread?

**Comment by kevlam2706** [comment]
Yup

Kevin

**Comment by ounsworth** [comment]
@George questions about Kafta: Since I've never used the technology, what would integrating with that from the Java code look like?
It looks like the installation and runtime is a bunch of shell scripts. Does Kafta ship binaries for ARM?  Do we know whether it's even legal to install software / modify boot scripts on the roboRIO?

**Comment by ounsworth** [comment]
As for roboRIO specs, it looks like it has 512 MB of disk and 256 MB RAM
http://www.ni.com/pdf/manuals/374661a.pdf

**Comment by georgetzavelas** [comment]
@ounsworth Kafka has a client library that we could use on the RoboRio. The server would be on the PC with the Dashboard. Just a thought if the students wanted to explore some technologies that are used to solve some of the problems we are seeing but there is a learning curve that may not be conducive to build season timing.
